{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MCB 288 - Spring 2019 - Exercise 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For week 2 we will be reanalyzing data from this paper:\n",
    "\n",
    "https://rnajournal.cshlp.org/content/22/6/839.long\n",
    "\n",
    "In which the authors did 96 RNA-seq experiments, 48 replicates each of wildtype yeast cells (S. cerevisiae) and an otherwise isogenic line carrying a mutation in the transcription factor SNF2. The main goal of the paper, which we will focus on here, was to ask the question of how many replicates are needed to accurately identify genes that are differentially expressed between the two conditions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have preprocessed the data a little bit to put it into a tabular form available in the file:\n",
    "\n",
    "##### data/Barton/Barton_combined_Ygenes.txt\n",
    "\n",
    "The file is a tab-delimited table whose rows are genes and columns are replicates. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the paper they identify several replicates that look unusual and which they do not include in subsequent analyses. We will discuss whether this was appropriate, but it is certainly true that several of the samples are unusual in that they look quite different from other supposedly identical samples."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Your assignment is to identify these outliers and describe what makes them unusual. You should be able to do this using only operations discussed in the last class. But I will start you off with some code to help."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries \n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# this is here to shut off some annoying warnings from pandas\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "# matplotlib is one of the main plotting libraries we're going to use\n",
    "import matplotlib \n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "# the other plotting library is seaborn - we'll use both during the class\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "# numpy and scipy are for handling numerical and scientific data\n",
    "\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "\n",
    "import statsmodels.formula.api as smf # basic statistical modeling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Snf2_rep01</th>\n",
       "      <th>Snf2_rep02</th>\n",
       "      <th>Snf2_rep03</th>\n",
       "      <th>Snf2_rep04</th>\n",
       "      <th>Snf2_rep05</th>\n",
       "      <th>Snf2_rep06</th>\n",
       "      <th>Snf2_rep07</th>\n",
       "      <th>Snf2_rep08</th>\n",
       "      <th>Snf2_rep09</th>\n",
       "      <th>...</th>\n",
       "      <th>WT_rep39</th>\n",
       "      <th>WT_rep40</th>\n",
       "      <th>WT_rep41</th>\n",
       "      <th>WT_rep42</th>\n",
       "      <th>WT_rep43</th>\n",
       "      <th>WT_rep44</th>\n",
       "      <th>WT_rep45</th>\n",
       "      <th>WT_rep46</th>\n",
       "      <th>WT_rep47</th>\n",
       "      <th>WT_rep48</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YAL001C</td>\n",
       "      <td>840</td>\n",
       "      <td>653</td>\n",
       "      <td>823</td>\n",
       "      <td>1036</td>\n",
       "      <td>510</td>\n",
       "      <td>820</td>\n",
       "      <td>1240</td>\n",
       "      <td>549</td>\n",
       "      <td>828</td>\n",
       "      <td>...</td>\n",
       "      <td>353</td>\n",
       "      <td>696</td>\n",
       "      <td>492</td>\n",
       "      <td>445</td>\n",
       "      <td>419</td>\n",
       "      <td>289</td>\n",
       "      <td>541</td>\n",
       "      <td>292</td>\n",
       "      <td>647</td>\n",
       "      <td>409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL002W</td>\n",
       "      <td>712</td>\n",
       "      <td>530</td>\n",
       "      <td>607</td>\n",
       "      <td>771</td>\n",
       "      <td>395</td>\n",
       "      <td>605</td>\n",
       "      <td>904</td>\n",
       "      <td>351</td>\n",
       "      <td>589</td>\n",
       "      <td>...</td>\n",
       "      <td>356</td>\n",
       "      <td>756</td>\n",
       "      <td>552</td>\n",
       "      <td>440</td>\n",
       "      <td>389</td>\n",
       "      <td>291</td>\n",
       "      <td>552</td>\n",
       "      <td>346</td>\n",
       "      <td>622</td>\n",
       "      <td>453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YAL003W</td>\n",
       "      <td>7296</td>\n",
       "      <td>6129</td>\n",
       "      <td>6464</td>\n",
       "      <td>7278</td>\n",
       "      <td>5613</td>\n",
       "      <td>3046</td>\n",
       "      <td>7687</td>\n",
       "      <td>6714</td>\n",
       "      <td>8382</td>\n",
       "      <td>...</td>\n",
       "      <td>9291</td>\n",
       "      <td>12055</td>\n",
       "      <td>13542</td>\n",
       "      <td>8441</td>\n",
       "      <td>7337</td>\n",
       "      <td>8639</td>\n",
       "      <td>14315</td>\n",
       "      <td>6618</td>\n",
       "      <td>14745</td>\n",
       "      <td>9878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YAL004W</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YAL005C</td>\n",
       "      <td>9851</td>\n",
       "      <td>10226</td>\n",
       "      <td>12006</td>\n",
       "      <td>10714</td>\n",
       "      <td>8575</td>\n",
       "      <td>2791</td>\n",
       "      <td>11764</td>\n",
       "      <td>8237</td>\n",
       "      <td>12213</td>\n",
       "      <td>...</td>\n",
       "      <td>10177</td>\n",
       "      <td>21176</td>\n",
       "      <td>13270</td>\n",
       "      <td>12337</td>\n",
       "      <td>14303</td>\n",
       "      <td>14268</td>\n",
       "      <td>14993</td>\n",
       "      <td>9607</td>\n",
       "      <td>16075</td>\n",
       "      <td>14797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene  Snf2_rep01  Snf2_rep02  Snf2_rep03  Snf2_rep04  Snf2_rep05  \\\n",
       "0  YAL001C         840         653         823        1036         510   \n",
       "1  YAL002W         712         530         607         771         395   \n",
       "2  YAL003W        7296        6129        6464        7278        5613   \n",
       "3  YAL004W           0           0           0           0           0   \n",
       "4  YAL005C        9851       10226       12006       10714        8575   \n",
       "\n",
       "   Snf2_rep06  Snf2_rep07  Snf2_rep08  Snf2_rep09    ...     WT_rep39  \\\n",
       "0         820        1240         549         828    ...          353   \n",
       "1         605         904         351         589    ...          356   \n",
       "2        3046        7687        6714        8382    ...         9291   \n",
       "3           0           0           0           0    ...            0   \n",
       "4        2791       11764        8237       12213    ...        10177   \n",
       "\n",
       "   WT_rep40  WT_rep41  WT_rep42  WT_rep43  WT_rep44  WT_rep45  WT_rep46  \\\n",
       "0       696       492       445       419       289       541       292   \n",
       "1       756       552       440       389       291       552       346   \n",
       "2     12055     13542      8441      7337      8639     14315      6618   \n",
       "3         0         0         0         0         0         0         0   \n",
       "4     21176     13270     12337     14303     14268     14993      9607   \n",
       "\n",
       "   WT_rep47  WT_rep48  \n",
       "0       647       409  \n",
       "1       622       453  \n",
       "2     14745      9878  \n",
       "3         0         0  \n",
       "4     16075     14797  \n",
       "\n",
       "[5 rows x 97 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load dataframe with expression data\n",
    "\n",
    "datafile = \"data/barton/Barton_combined_Ygenes.txt\"\n",
    "df = pd.read_csv(datafile, sep='\\t')   # the sep='\\t' tells pandas that it is a tab separated file\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not that the first column is \"Gene\" and uses the standard identifier for yeast genes which has several components:\n",
    "\n",
    "Take the first one as an example: YAL001C\n",
    "\n",
    "* Y - is for yeast\n",
    "* A - tells you what chromosome it is on (A is chromosome 1, B is chromosome 2, etc...). Yeast has 16 chromosomes plus mitcochondria.\n",
    "* L - is for the left arm of the chromosome (the other arm is R for right)\n",
    "* 001 - is a numerical identifier counting genes along the chromosome arm\n",
    "* C - says that it is on the \"Crick\" strand - the other option is W for \"Watson\" (we should all get together and lobby to have this changed to F for Franklin)\n",
    "\n",
    "You can get all sorts of info on yeast genes at the Yeast Genome Database www.yeastgenome.org.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# you will likely want to use lists of experiments \n",
    "\n",
    "# this comment gets all of the column names and puts them into a list\n",
    "\n",
    "exps = df.columns.get_values()\n",
    "\n",
    "# you can then create lists with just WT and Snf2 mutants\n",
    "\n",
    "# the syntax here in English is \n",
    "# [] create a new list\n",
    "# [e for e in exps] fills the list with all values from the list exps\n",
    "# if e.startswith('WT') filters the list to only include elements that start with 'WT'\n",
    "\n",
    "exps_wt = [e for e in exps if e.startswith('WT')] \n",
    "\n",
    "exps_mut = [e for e in exps if e.startswith('Snf2')] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 WT_rep01 891.6344053851907\n",
      "1 WT_rep02 1179.07928197457\n",
      "2 WT_rep03 961.8833208676141\n",
      "3 WT_rep04 1558.8936424831713\n",
      "4 WT_rep05 1197.2359012715033\n",
      "5 WT_rep06 2184.868511593119\n",
      "6 WT_rep07 1324.374569932685\n",
      "7 WT_rep08 1471.038444278235\n",
      "8 WT_rep09 1323.5594614809274\n",
      "9 WT_rep10 1183.7739715781602\n",
      "10 WT_rep11 1015.8858638743455\n",
      "11 WT_rep12 1227.67928197457\n",
      "12 WT_rep13 1357.6833208676142\n",
      "13 WT_rep14 1435.7350785340313\n",
      "14 WT_rep15 1138.921166791324\n",
      "15 WT_rep16 831.9109947643979\n",
      "16 WT_rep17 1073.470306656694\n",
      "17 WT_rep18 1149.562004487659\n",
      "18 WT_rep19 1394.1862378459236\n",
      "19 WT_rep20 939.7748691099476\n",
      "20 WT_rep21 1059.503216155572\n",
      "21 WT_rep22 1527.439192221391\n",
      "22 WT_rep23 1141.7406133133882\n",
      "23 WT_rep24 1188.7741211667912\n",
      "24 WT_rep25 1229.777860882573\n",
      "25 WT_rep26 1379.4103216155572\n",
      "26 WT_rep27 1433.9470456245326\n",
      "27 WT_rep28 1204.9184741959612\n",
      "28 WT_rep29 1219.3177262528047\n",
      "29 WT_rep30 1727.1461480927449\n",
      "30 WT_rep31 1286.499177262528\n",
      "31 WT_rep32 977.5537771129394\n",
      "32 WT_rep33 1420.2958863126403\n",
      "33 WT_rep34 1277.6635751682873\n",
      "34 WT_rep35 994.6737471952132\n",
      "35 WT_rep36 1172.9449513836948\n",
      "36 WT_rep37 1646.1175766641736\n",
      "37 WT_rep38 1076.1056095736724\n",
      "38 WT_rep39 1107.6121166791324\n",
      "39 WT_rep40 1782.4139117427076\n",
      "40 WT_rep41 1490.9669409124906\n",
      "41 WT_rep42 1094.0667165295438\n",
      "42 WT_rep43 990.1554225878833\n",
      "43 WT_rep44 1212.076439790576\n",
      "44 WT_rep45 1730.5044128646223\n",
      "45 WT_rep46 834.1238593866866\n",
      "46 WT_rep47 1805.9500373971578\n",
      "47 WT_rep48 1195.2303664921467\n"
     ]
    }
   ],
   "source": [
    "# there are a lot of ways you could now ask the question\n",
    "# \"How do the different WT replicates compare to each other?\"\n",
    "# I will leave you with some code to help get you started\n",
    "\n",
    "\n",
    "# this command iterates over the elements in a list returning \n",
    "# both the index of the element - 0,1,2,3... - which i'm calling i\n",
    "# and the actual value which i'm calling e\n",
    "\n",
    "for i,e in enumerate(exps_wt):\n",
    "    \n",
    "    # you can then use these values to do stuff\n",
    "    \n",
    "    print (i,e,df[e].mean())\n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if you want to compare all WT experiments to each other you could do something like this\n",
    "\n",
    "# import the pearson correlation function\n",
    "from scipy.stats.stats import pearsonr \n",
    "\n",
    "for i1,e1 in enumerate(exps_wt):\n",
    "    for i2,e2 in enumerate(exps_wt):\n",
    "        c = pearsonr(df[e1],df[e2])[0] \n",
    "        \n",
    "# this calculates the pearson correlation for all pairs of WT experiments\n",
    "# I will leave it up to you to figure out what to do with it"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
